BinaryEncoding.md

```diff
@@ -51,19 +51,26 @@ A single-byte unsigned integer indicating a [value type](AstSemantics.md#types).
 
 # Definitions
 
-### Pre-order encoding
-Refers to an approach for encoding syntax trees, where each node begins with an identifying binary
-sequence, then followed recursively by any child nodes. 
-
-* Examples
-  * Given a simple AST node: `I32Add(left: AstNode, right: AstNode)`
-    * First write the opcode for `I32Add` (uint8)
-    * Then recursively write the left and right nodes.
-
-  * Given a call AST node: `Call(callee_index: uint32_t, args: AstNode[])`
-    * First write the opcode of `Call` (uint8)
-    * Then write the (variable-length) integer `callee_index` (varuint32)
-    * Then recursively write each argument node, where arity is determined by looking up `callee_index` in a table of signatures
+### "Post-order" syntax tree encoding
+
+WASM syntax trees are encoded in a variation of post-order traversal. Nodes with fixed arity will be encoded in post-order. 
+Nodes with variable-arity such as but not limited to blocks, loops, and calls will be delimited with opcode-specific start and end markers. Thus, for fixed arity, the binary sequence begins with child subtrees followed by the opcode.
+For variable arity, the binary sequence begins with an opcode-specific start marker, followed by the the child subtrees (encoded in this same variant of post-order), then the opcode, which serves as an end marker.
+This encoding is immediately amenable to decoding with an explicit stack without need for shift-reduce parsing. 
+
+####Examples:
+
+- Given a simple AST node: I32Add(left: AstNode, right: AstNode)
+  - First recursively write the left and right nodes
+  - Then write the opcode for I32Add, which is sufficient since I32Add is a binary opcode.   
+
+- Given nested call AST nodes: Call(0, [Call(1, [I32Const 2, I32Const 3]), I32Const 4])
+  - First, write the opcode marker CALL_START
+  - Recursively write the first operand of the outer Call: 0 
+  - Recursively write the second operand: CALL_START 1 2 I32CONST 3 I32CONST CALL_END
+  - Recursively the write third operand: 4 I32CONST
+  - Write the opcode CALL_END 
```

Comment:
Could some rationale be given for the need to have a `call_start` marker? I can see the need to know the start of control flow forks, for the single-pass SSA conversion, and for blocks to know the break depths and perhaps other reasons, but no need comes to mind for calls and it seems adequate to just post-order encode them?